### PR TITLE
Use string_id for mtype ids

### PR DESCRIPTION
**Pull request by BevapDin**
As the title says, this uses the type `mtype_id` (which is `string_id<mtype>`) for mtype ids.

Because literal ids are used quite often in some files, I extracted those literals into (global) static, const objects. That is, I was simply to lazy to add `mtype( )` everywhere (changing `bay.add_spawn( "mon_zombie_electric", ...` to `bay.add_spawn( mtype_id( "mon_zombie_electric" ), ...`).

Also gets rid of the `GetMType` macro. One can either use the `MonsterGenerator` directly or use `mtype_id::obj()`.

Some functions that used to take or return a `mtype` pointer now use `mtype_id` instead (`get_valid_hallucination`).

The commit messages explain the changes.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Most hardcoded mon types are referred to by constants. Is it now expensive to look types up?

**Comment by BevapDin** [comment]
> Most hardcoded mon types are referred to by constants. Is it now expensive to look types up?

No, the look up is not changed, but the constants are easier to manage and are a tiny bit faster.

It's still possible (and valid) to do
```C++
monster mon( mtype_id( "mon_dog" ) );
```
It's a bit more verbose (the `mtype(  )`), which I got rid by moving everything into the constants.

**Comment by Coolthulhu** [comment]
Oops, forgot to run `make tests` - savegame_test.cpp is now broken. Easily fixable (to get it to compile), but I couldn't test if the fix actually fixes anything beyond compilation.

File: src/cata_tiles.cpp [hunk 1]
```diff
@@ -1774,38 +1776,30 @@ void cata_tiles::tile_loading_report(maptype const & tiletypemap, std::string co
...
-        const std::string &s = sid.str();
-        if (tile_ids.count(prefix+s) == 0) {
-            missing++;
-            missing_list.append(s+" ");
-        } else {
-            present++;
-        }
-    }
-    DebugLog( D_INFO, DC_ALL ) << "Missing " << label << ": " << missing_list;
+    lr_generic( 0u, count,
```
Comment (narc0tiq) [inline src/cata_tiles.cpp]: Windows_x64 tiles build fails:
```
src/cata_tiles.cpp:1763:6: note:   template argument deduction/substitution failed:
src/cata_tiles.cpp:1790:5: note:   deduced conflicting types for parameter 'Iter' ('unsigned int' and 'long long unsigned int')
     lr_generic( 0u, count,
     ^
```
-- via <http://ci.narc.ro/job/Cataclysm-Matrix/Graphics=Tiles,Platform=Windows_x64/3494/console>

Probably wanted something like `lr_generic( (size_t)0, count,`?

**Pull request merged**